### PR TITLE
add support for the UploadQuotePolicy and GetQuotePolicy restful API

### PR DESCRIPTION
**Pull request by wanghouqi**
1. these restful API will be exposed to public as:
 POST <ehsm_srv_address>/ehsm?Action=
  UploadQuotePolicy: upload the value of mrenclave and mrsigner and return the policy ID to user.
  GetQuotePolicy: get the policy of mrencalve and mrsigner by policy ID.

2. Modify VerifyQuote API to support policy verification

test: passed the python test

Signed-off-by: wanghouqi <wanghouqi@vip.qq.com>

### CODE REVIEW

File: core/App/ehsm_core_test.cpp [hunk 1]
```diff
@@ -956,11 +956,14 @@ void test_GenerateQuote_and_VerifyQuote()
     printf("============test_GenerateQuote_and_VerifyQuote start==========\n");
     char challenge[32] = "challenge123456";
     char nonce[16] = "nonce123456";
-
+    // the string generated after converting the value of mr_signer and mr_enclave to hexadecimal
+    // notice: these 2 values will be changed if our enclave has been updated. then the case may be failed.
```
Comment (syan10) [inline core/App/ehsm_core_test.cpp:960]: you can add the cmd that i sent to you to dump the mrenclave/mrsigner as comments.

File: core/App/ehsm_provider.cpp [hunk 2]
```diff
@@ -1013,6 +1029,17 @@ ehsm_status_t VerifyQuote(ehsm_data_t *quote,
     //set current time. This is only for sample use, please use trusted time in product.
     current_time = time(NULL);
 
+    // check mr_signer and mr_enclave
+    if((mr_signer != NULL &&  strncmp(mr_signer, " ", strlen(mr_signer)) != 0) || 
+       (mr_enclave != NULL && strncmp(mr_enclave, " ", strlen(mr_enclave)) != 0)) {
+        ret = enclave_verify_policy(g_enclave_id, &sgxStatus, quote->data, quote->datalen, mr_signer, strlen(mr_signer), mr_enclave, strlen(mr_enclave));
```
Comment (syan10) [inline core/App/ehsm_provider.cpp:1035]: follow the line#1044 to format the function parameters.

File: core/Enclave/enclave_hsm.edl [hunk 3]
```diff
@@ -129,6 +129,10 @@ enclave {
         public sgx_status_t enclave_get_rand([out, size=datalen] uint8_t *data,
                                                 uint32_t datalen);
 
+        public sgx_status_t enclave_verify_policy([in, size=quote_len] uint8_t* quote, uint32_t quote_len, 
+                            [in, size=mr_signer_good_len] const char* mr_signer_good, uint32_t mr_signer_good_len,
```
Comment (syan10) [inline core/Enclave/enclave_hsm.edl:133]: please keep the same format with other API

File: core/App/ehsm_provider.cpp (hunk 2 above)
Comment (houhuayou) [reply core/App/ehsm_provider.cpp:1035]: Ok, I will format it.

File: core/App/ehsm_core_test.cpp (hunk 1 above)
Comment (houhuayou) [reply core/App/ehsm_core_test.cpp:960]: Ok, I'll add it.

File: core/Enclave/enclave_hsm.edl (hunk 3 above)
Comment (houhuayou) [reply core/Enclave/enclave_hsm.edl:133]: I don't understand what you mean. Could you say it in more detail

Comment (syan10) [reply core/Enclave/enclave_hsm.edl:133]: keep alignment

Comment (houhuayou) [reply core/Enclave/enclave_hsm.edl:133]: Ok, I will change it.

**Review by syan10 — APPROVED** [review]

File: core/Enclave/enclave_hsm.cpp [hunk 4]
```diff
@@ -976,4 +979,42 @@ sgx_status_t enclave_verify_att_result_mac(sgx_ra_context_t context,
...
+    char mr_signer_temp[3] = {0};
+    char mr_enclave_temp[3] = {0};
+    sgx_quote3_t *p_sgx_quote = (sgx_quote3_t *)quote;
+    for(int i = 0; i < SGX_HASH_SIZE; i++) {
+        snprintf(mr_signer_temp, sizeof(mr_signer_temp) , "%02x", p_sgx_quote->report_body.mr_signer.m[i]);
+        snprintf(mr_enclave_temp, sizeof(mr_enclave_temp), "%02x", p_sgx_quote->report_body.mr_enclave.m[i]);
+        mr_signer_str += mr_signer_temp;
+        mr_enclave_str += mr_enclave_temp;
+    }
+    if(strncmp(mr_signer_good, mr_signer_str.c_str(), mr_signer_str.size()) != 0 ||
```
Comment (yang8621) [inline core/Enclave/enclave_hsm.cpp:1014]: the length of mr_signer_good and mr_signer_str is not checked.

Comment (yang8621) [reply core/Enclave/enclave_hsm.cpp:1014]: > the length of mr_signer_good and mr_signer_str is not checked. The two should be the same.

E.g. The known good value is 1234567890, while a quote is 12345.

Comment (houhuayou) [reply core/Enclave/enclave_hsm.cpp:1014]: OK, I will compare their lengths before comparing their content.

**Review by yang8621 — APPROVED** [review]

**Review by HawieLong — APPROVED** [review]

**Pull request merged**